feat: make Docker image self-contained

jkuester opened the pull request:
The goal of this PR is to further operationalize this service with the goal of making it easier to consume (without limiting its flexibility).

I have updated the Dockerfile to include the necessary configuration files and default envar values in the Docker image, itself, (instead of needing to inject them at runtime).  This allows the image to be used without needing to clone the nginx-local-ip repo.

If this approach is acceptable, I have also included GitHub action configuration which should allow for automatically publishing a new version of the docker image whenever new changes are pushed to the `main` branch of this repo.  This config is based on what we are using [in cht-conf](https://github.com/medic/cht-conf/blob/main/.github/workflows/publish_docker_image.yml) to publish the cht-app-ide image.  Some additional manual configuration will be needed to get the image building to work (which I can do if this PR is approved).  Context:
- https://github.com/medic/cht-conf/pull/546
- https://github.com/medic/medic-infrastructure/issues/570
- https://github.com/medic/cht-conf/pull/558
- https://github.com/medic/cht-conf/pull/559

jkuester commented on Dockerfile:
```diff
@@ -1,6 +1,13 @@
-FROM nginx:1.19
+FROM nginx:1.28-alpine
```
Changing to the `alpine` image is not strictly necessary for these changes, but it image is 2-3x smaller than the default image....

jkuester commented on compose.yaml:
```diff
@@ -10,7 +8,6 @@ services:
       - "${HTTPS}:443"
     volumes:
       - ./default.conf.template:/etc/nginx/templates/default.conf.template
-      - ./entrypoint.sh:/entrypoint.sh
```
Seems like it is useful to show how to override the `default.conf.template` since consumers may want to customize that.  But, I don't think we need to demonstrate overriding the `entrypoint.sh`.  If you need a custom version of that script, you should maybe consider building a different image.... :thinking:

mrjones-plip reviewed (changes requested):
yah!  This is a great revival of this project - thanks Josh!

I have no concerns with any of the updates - they all make sense.  We'll have to work out some of image publishing stuffs as always (new ECR repo? adding GH CI secrets? etc).  We'll get to those as needed!

I did see an error in the entrypoint that should be tended to, `entrypoint.sh: line 41: openssl: not found ` and  `grep: unrecognized option: P` and see below

As well, and I doubt we can fix this, the handy "this is your URL!!" gets pushed off the screen.  be rad if we could have it output very last. likely not though, in which case we should very clearly call out in docs to "be sure to scroll up" or something ;)

### demo of lost  "this is your URL!!"  in entry point

https://github.com/user-attachments/assets/639695d9-0a78-4b7d-9259-433845565b1b


### entrypoint error

```
nginx-local-ip  | /entrypoint.sh: Downloading 'https://local-ip.medicmobile.org/chain' ...                                     
nginx-local-ip  | /entrypoint.sh: Creating chained cert file '/etc/nginx/server.chained.pem' ...                               
nginx-local-ip  | /entrypoint.sh: line 41: openssl: not found                                                                  
nginx-local-ip  | grep: unrecognized option: P                                                                                 
nginx-local-ip  | BusyBox v1.37.0 (2025-01-17 18:12:01 UTC) multi-call binary.                                                 
nginx-local-ip  |                                                                                                              
nginx-local-ip  | Usage: grep [-HhnlLoqvsrRiwFE] [-m N] [-A|B|C N] { PATTERN | -e PATTERN... | -f FILE... } [FILE]...          
nginx-local-ip  |                                                                                                              
nginx-local-ip  | Search for PATTERN in FILEs (or stdin)                                                                       
nginx-local-ip  |                                                                                                              
nginx-local-ip  |       -H      Add 'filename:' prefix                                                                         
nginx-local-ip  |       -h      Do not add 'filename:' prefix                                                                  
nginx-local-ip  |       -n      Add 'line_no:' prefix                                                                          
nginx-local-ip  |       -l      Show only names of files that match                                                            
nginx-local-ip  |       -L      Show only names of files that don't match                                                      
nginx-local-ip  |       -c      Show only count of matching lines                                                              
nginx-local-ip  |       -o      Show only the matching part of line                                                            
nginx-local-ip  |       -q      Quiet. Return 0 if PATTERN is found, 1 otherwise                                               
nginx-local-ip  |       -v      Select non-matching lines                                                                      
nginx-local-ip  |       -s      Suppress open and read errors                                                                  
nginx-local-ip  |       -r      Recurse                                                                                        
nginx-local-ip  |       -R      Recurse and dereference symlinks                                                               
nginx-local-ip  |       -i      Ignore case                                                                                    
nginx-local-ip  |       -w      Match whole words only                                                                         
nginx-local-ip  |       -x      Match whole lines only                                                                         
nginx-local-ip  |       -F      PATTERN is a literal (not regexp)                                                              
nginx-local-ip  |       -E      PATTERN is an extended regexp                                                                  
nginx-local-ip  |       -m N    Match up to N times per file                                                                   
nginx-local-ip  |       -A N    Print N lines of trailing context                                                              
nginx-local-ip  |       -B N    Print N lines of leading context                                                               
nginx-local-ip  |       -C N    Same as '-A N -B N'                                                                            
nginx-local-ip  |       -e PTRN Pattern to match                                                                               
nginx-local-ip  |       -f FILE Read pattern from file                                                                         
nginx-local-ip  | date: invalid date ''                                                                                        
nginx-local-ip  | /entrypoint.sh: SSL certificate OK. Expire after:                                                            
nginx-local-ip  | /entrypoint.sh: line 45: can't open 2025-05-27: no such file                                                 
nginx-local-ip  | /entrypoint.sh:                                                                                              
nginx-local-ip  |     --------------------------------------   
nginx-local-ip  |     nginx-local-ip URL:                      
nginx-local-ip  |     https://192-168-0-3.local-ip.medicmobile.org                                                             
nginx-local-ip  |     --------------------------------------   
nginx-local-ip  | /entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration                    
nginx-local-ip  | /entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/                                           
nginx-local-ip  | /entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh  
```

jkuester commented:
Thanks for the review @mrjones-plip! Looks like the script errors were caused by the `Alpine` environment.  Made some minor changes and things seem to be working as desired now:

- `openssl: not found` - Just installed `openssl` in Dockerfile
- `grep: unrecognized option: P` - Perl-style regex is not supported in the Alpine `grep`.  Switched to use `sed` instead.
- `date: invalid date ''` - Alpine's base `date` util is super limited.  Installed `coreutils` in Dockerfile to get the GNU `date` util instead.  Also updated the `if` logic for checking the EXP_DATE to be POSIX compliant (via just comparing the epoch number values) instead of using bash-only features.

> the handy "this is your URL!!" gets pushed off the screen

Unfortunately, I could not figure out a simple way to do this.  We need to call `. /docker-entrypoint.sh` last in the script and that it what starts outputting the additional content (and we want to also see this output because this is the nginx logs...).  So, I have added a note to the README to scroll up to see your URL.  :+1:

jkuester commented:
Hey @mrsarm, great to hear from you!

That is a good call on the worker processes. :thinking: I did a bit of checking and it seems a bit non-trivial to tune the worker process count on nginx running in the docker container.  E.g. see: https://github.com/nginx/docker-nginx/issues/472.

I did confirm that we seems to currently be running with `worker_process auto;` (which probably does not respect any container-level CPU provisioning.  Perhaps we need to set `NGINX_ENTRYPOINT_WORKER_PROCESSES_AUTOTUNE`?  Alternatively, we could try to hard-code the a lower number of worker processes since (as you pointed out) we are not too worried about load here....  Either way, I would rather push that into a separate issue and not try to solve it in this PR.  :sweat:

mrjones-plip reviewed (approved):
yeah - looks good! Thanks for tending to those errors and investigating initial boot output (and hello @mrsarm :wave: !!)

I think this is good to go but calling out that if you have a local image built already (which 100% of current users do), then you get an error when you run the `docker run` command blindly.  Maybe when we publish a newer image this will fix itself?  My work around was to delete the image (`docker image rm -f medicmobile/nginx-local-ip`) , build the image with the docker compose command (`docker compose up`), then it worked as expected

Here's the error for reference:

```
➜  nginx-local-ip git:(proper-docker-image) docker run --rm \
    -e APP_URL=http://192.168.0.3:5988 \
    -p 443:443 \
    medicmobile/nginx-local-ip
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: exec: "/entrypoint.sh": stat /entrypoint.sh: no such file or directory: unknown

Run 'docker run --help' for more information
```
